### PR TITLE
304 travis local deploy

### PR DESCRIPTION
**Pull request by slifty**
This PR adds support for browser tests in travis via SauceLabs.  It requires Travis to be configured with SauceLabs credentials.  Currently the browser tested is Chrome on OSX, but this can be replaced with any platform supported on Sauce Labs.

Part of the PR is the addition of two new scripts:

1. A script for setting up wildfly on travis.
2. A script for deploying the compile application to wildfly on travis.

They appear to be consistently running; we should keep an eye out for resource errors associated with wildfly as this was an issue in earlier iterations of the script.  It appears that by starting the server during the `before_script` block we avoid race conditions with build and deploy.

### CODE REVIEW

**Comment by slifty** [comment]
I failed to mention this in the PR notes: I needed to upgrade Serenity because there seemed to be SauceLabs compatibility issues in the version we were using.

File: .travis.yml [hunk 1]
```diff
@@ -27,6 +33,10 @@ script:
         cms-portal-services:build
         cms-web:apiDocs
         userhelp:html
+  - ../scripts/travis-deploy-application.sh
+  - ./gradlew --no-daemon --console=plain
+        test
+        aggregate
```
Comment (jasonaowen) [inline .travis.yml]: The `aggregate` task builds a bunch of HTML; if we're not doing anything with that, and it gets thrown away with the rest of the worker VM, we should skip it.

File: psm-app/build.gradle [hunk 2]
```diff
@@ -270,8 +270,8 @@ project(':integration-tests') {
     }
 
     dependencies {
-        testCompile 'net.serenity-bdd:serenity-core:1.4.0'
-        testCompile 'net.serenity-bdd:serenity-junit:1.4.0'
+        testCompile 'net.serenity-bdd:serenity-core:1.5.5'
+        testCompile 'net.serenity-bdd:serenity-junit:1.5.5'
```
Comment (jasonaowen) [inline psm-app/build.gradle]: It looks like these two should be upgraded in sync (right?), so would you mind extracting a variable for the version number? And, could you please separate this change into its own commit with an explanation of why?

File: psm-app/integration-tests/src/test/java/gov/medicaid/features/enrollment/ui/LicenseInfoPage.java [hunk 3]
```diff
@@ -30,14 +30,10 @@ public void uploadSampleFile() throws IOException {
                 new FileOutputStream(temporaryFile))) {
             writer.println("License");
         }
-        uploadFile(temporaryFile.getAbsolutePath());
+        upload(temporaryFile.getAbsolutePath()).to($("[name=_03_attachment_0]"));
```
Comment (jasonaowen) [inline psm-app/integration-tests/src/test/java/gov/medicaid/features/enrollment/ui/LicenseInfoPage.java]: This is great! Can you please extract this into its own commit explaining why?

File: scripts/travis-deploy-application.sh [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+WILDFLY_HOME=${TRAVIS_BUILD_DIR}/../wildfly-10.1.0.Final
+WILDFLY_CLI=${WILDFLY_HOME}/bin/jboss-cli.sh
+WILDFLY_LOG=$(mktemp --tmpdir wildfly.XXXXXX.log)
+
+function download_and_sha1 {
+  base=$(basename $1)
+  [ -e $base ] || curl -OL $1
+  echo "$2 $base" | sha1sum --quiet -c - \
+    || (printf "$base has an unexpected sha1sum: $(sha1sum $base | awk '{print $1}')\n" && exit)
+}
```
Comment (jasonaowen) [inline scripts/travis-deploy-application.sh]: I think these are the third and fourth times we've copy-pasted the `download_and_sha1` function; I think it's time to extract a `source`-able script!

File: scripts/travis-start-wildfly.sh [hunk 5]
```diff
@@ -22,7 +22,9 @@ function wait_for_wildfly {
 
 createdb -U postgres psm
 
-${WILDFLY_HOME}/bin/standalone.sh -c standalone-full.xml > ${WILDFLY_LOG} &
+${WILDFLY_HOME}/bin/standalone.sh -c standalone-full.xml -b 0.0.0.0 > ${WILDFLY_LOG} &
```
Comment (jasonaowen) [inline scripts/travis-start-wildfly.sh]: Why make the management console available?

File: scripts/travis-start-wildfly.sh [hunk 6]
```diff
@@ -67,17 +69,12 @@ xa-data-source add \
   --password=ignored \
   --xa-datasource-properties=ServerName=localhost,PortNumber=5432,DatabaseName=psm
 EOF
-${WILDFLY_CLI} --connect --command="reload"
-${WILDFLY_CLI} --connect --command="xa-data-source test-connection-in-pool --name=MitaDS"
-${WILDFLY_CLI} --connect --command="xa-data-source test-connection-in-pool --name=TaskServiceDS"
-${WILDFLY_CLI} --connect --command="deploy ${TRAVIS_BUILD_DIR}/psm-app/cms-portal-services/build/libs/cms-portal-services.ear"
-
-python -m smtpd -n -c DebuggingServer localhost:1025 &
```
Comment (jasonaowen) [inline scripts/travis-start-wildfly.sh]: Why not include the test smtpd?

File: psm-app/build.gradle [hunk 7]
```diff
@@ -289,8 +289,14 @@ project(':integration-tests') {
...
-        environment "DISPLAY", System.getenv('XVFB_DISPLAY')
-        systemProperty "webdriver.firefox.bin", System.getenv('FIREFOX_HOME')
+        // TODO: Platform, driver, and version can be env configurable
+        systemProperty "webdriver.driver",  "chrome"
+        systemProperty "saucelabs.url", "http://" + System.getenv("SAUCE_USERNAME") + ":" + System.getenv("SAUCE_ACCESS_KEY") + "@ondemand.saucelabs.com:80/wd/hub"
+        systemProperty "saucelabs.target.platform", "MAC" // See: https://github.com/SeleniumHQ/selenium/blob/master/java/client/src/org/openqa/selenium/Platform.java
+        systemProperty "saucelabs.driver.version", "60"
+        systemProperty "saucelabs.access.key", System.getenv("SAUCE_ACCESS_KEY")
+        systemProperty "saucelabs.user.id", System.getenv("SAUCE_USERNAME")
+        systemProperty "saucelabs.tunnel-identifier", System.getenv("TRAVIS_JOB_NUMBER")
```
Comment (jasonaowen) [inline psm-app/build.gradle]: So... how does this work? And how does it affect local testing? Can we put it in a separate gradle task, like `travisSaucelabsTest` or something?

File: psm-app/build.gradle [hunk 8]
```diff
@@ -289,8 +289,14 @@ project(':integration-tests') {
     }
 
     test {
-        environment "DISPLAY", System.getenv('XVFB_DISPLAY')
-        systemProperty "webdriver.firefox.bin", System.getenv('FIREFOX_HOME')
+        // TODO: Platform, driver, and version can be env configurable
+        systemProperty "webdriver.driver",  "chrome"
+        systemProperty "saucelabs.url", "http://" + System.getenv("SAUCE_USERNAME") + ":" + System.getenv("SAUCE_ACCESS_KEY") + "@ondemand.saucelabs.com:80/wd/hub"
+        systemProperty "saucelabs.target.platform", "MAC" // See: https://github.com/SeleniumHQ/selenium/blob/master/java/client/src/org/openqa/selenium/Platform.java
```
Comment (cecilia-donnelly) [inline psm-app/build.gradle]: Why Mac?  (I skimmed the link and it looks like `LINUX` is also available -- is this a Sauce Labs limitation?)

File: psm-app/build.gradle (hunk 7 above)
Comment (slifty) [reply psm-app/build.gradle]: Sure thing -- I was under the impression that the test block is kind of dead code right now since the instructions for testing involves changing / deleting it; I can break it out, or we could add a conditional to the core test block (or both!)

For now I'll break it out and call the separate test block directly.

File: scripts/travis-start-wildfly.sh (hunk 5 above)
Comment (slifty) [reply scripts/travis-start-wildfly.sh]: I think this was a byproduct of debugging, good catch.  I suspect removing it will have no negative impact.

File: scripts/travis-start-wildfly.sh (hunk 6 above)
Comment (slifty) [reply scripts/travis-start-wildfly.sh]: What value does it provide?

File: psm-app/build.gradle (hunk 8 above)
Comment (slifty) [reply psm-app/build.gradle]: I think if we can only test one that Mac or Windows is the best choice since I doubt many of the application's users are going to be using Linux -- keep in mind this setting is for the front end tests (but if my assumption is incorrect and we want Linux, let me know).

I started with Windows and switched to Mac because there were some file system issues (which I suspect are no longer issues now that we are using the proper upload code for serenity)

Ultimately we should be testing on all of them eventually.

File: psm-app/build.gradle (hunk 7 above)
Comment (jasonaowen) [reply psm-app/build.gradle]: It is, in that -- as you say -- we've been deleting it to run the tests locally, which is less than ideal. Will we also have to delete the new version to run the tests locally?

File: scripts/travis-start-wildfly.sh (hunk 6 above)
Comment (jasonaowen) [reply scripts/travis-start-wildfly.sh]: It allows the application to send emails (as far as it knows, anyway...) which will be important if and when we add any automated tests that cover any of the email use cases. On the other hand, YAGNI... but it's easy to imagine adding a test for such a use case, getting it working locally, and then being confounded by Travis showing them errors.

File: psm-app/build.gradle (hunk 8 above)
Comment (cecilia-donnelly) [reply psm-app/build.gradle]: Ah that makes sense.  Thanks!  And yes to testing on all of them, but Mac seems like a fine place to start.

File: psm-app/build.gradle [hunk 9]
```diff
@@ -289,8 +289,14 @@ project(':integration-tests') {
     }
 
     test {
-        environment "DISPLAY", System.getenv('XVFB_DISPLAY')
-        systemProperty "webdriver.firefox.bin", System.getenv('FIREFOX_HOME')
+        // TODO: Platform, driver, and version can be env configurable
+        systemProperty "webdriver.driver",  "chrome"
+        systemProperty "saucelabs.url", "http://" + System.getenv("SAUCE_USERNAME") + ":" + System.getenv("SAUCE_ACCESS_KEY") + "@ondemand.saucelabs.com:80/wd/hub"
```
[the remark this answers is not in the document]
Comment (slifty) [reply psm-app/build.gradle]: (good catch; should probably be https.  That was from copypasta I think)

File: psm-app/build.gradle (hunk 7 above)
Comment (slifty) [reply psm-app/build.gradle]: I imagine so, unless you set your local environment to work with sauce labs; my goal with this PR wasn't to fix the local test workflow so much as it was to make it work on travis.

File: psm-app/integration-tests/src/test/java/gov/medicaid/features/enrollment/ui/LicenseInfoPage.java (hunk 3 above)
Comment (cecilia-donnelly) [reply psm-app/integration-tests/src/test/java/gov/medicaid/features/enrollment/ui/LicenseInfoPage.java]: Maybe @slifty can just add a comment here?  Sorry I didn't respond to this earlier, @jasonaowen, but this seems like part of the same conceptual change.  Personally, I think a comment here would be just as helpful as a different entry in the commit history, if not more so.  Was there a particular reason why you wanted to separate this out, though?

File: scripts/travis-deploy-application.sh (hunk 4 above)
Comment (slifty) [reply scripts/travis-deploy-application.sh]: Seems reasonable, how about we make an issue for it and a separate PR -- and we can clean this one up at the same time (rather than touching every place that used the code in this PR)

Issue here: https://github.com/OpenTechStrategies/psm/issues/495

File: psm-app/build.gradle (hunk 7 above)
Comment (slifty) [reply psm-app/build.gradle]: Actually sorry, those blocks I think are specifically named gradle blocks, so we can't just rename them and expect them to hook.  I will wrap it in the travis check conditional.

File: scripts/travis-start-wildfly.sh (hunk 6 above)
Comment (slifty) [reply scripts/travis-start-wildfly.sh]: I shall put it back!

File: psm-app/integration-tests/src/test/java/gov/medicaid/features/enrollment/ui/LicenseInfoPage.java (hunk 3 above)
Comment (slifty) [reply psm-app/integration-tests/src/test/java/gov/medicaid/features/enrollment/ui/LicenseInfoPage.java]: I'm happy to break it out but am also inclined to think of this as part of the same conceptual change; for now I definitely will at least amend the commit message to explain)

This change was put in because Serenity has its own built in upload method which supports remote browsers being controlled from a local machine.  This is important because not only does the file being uploaded not exist on the remote machine, but that machine might use an entirely different file system (meaning absolute paths are no longer absolute depending on where you are running / running from).

File: docs/TESTING.md [hunk 10]
```diff
@@ -38,6 +38,25 @@ buttons in the `ui` subdirectory.
...
+#### With SauceLabs
+Sauce Labs is a proprietary service that is free for open source projects.
+We use it for our continuous integration, but local developers are welcome
+to use it for local tests as well.  This documentation is geared towards
+setting up continuous integration, but the environment variables are the
+same in both cases.  To set up local sauce labs testing, please read the
+[sauce labs documentation](https://wiki.saucelabs.com/display/DOCS/Sauce+Connect+Proxy).
+
+**Environment Variables:**
+These must be set in the testing environment in order to communicate with
```
Comment (cecilia-donnelly) [inline docs/TESTING.md]: The testing environment is Travis, in this case?

Comment (slifty) [reply docs/TESTING.md]: Yes -- will clear up

**Comment by cecilia-donnelly** [comment]
So @slifty pointed out that `travis-ssh-add.sh` is not empty after all -- hurray!

I was able to build, deploy, and click around in the app locally.  When I ran the Selenium tests, two of them failed:
```
gov.medicaid.features.IntegrationTests > Scenario: Captures Individual Beneficial Owner's City.When I click 'next' on the Ownership Info Page FAILED
    org.openqa.selenium.WebDriverException

gov.medicaid.features.IntegrationTests > Scenario: Captures Individual Beneficial Owner's City.classMethod FAILED
    org.openqa.selenium.WebDriverException
```

Checking now to see if that's true on master, too.

**Comment by slifty** [comment]
@cecilia-donnelly The tests do all pass on Travis (which is the focus of this branch).

For local tests to pass you may need to configure your local copy to run tests properly; it's also possible you're running a different version of the application than what the tests have been written against.  The fact that tests pass in a controlled environment is a hopeful indicator that the branch itself is in good shape.

**Comment by cecilia-donnelly** [comment]
After discussion with @slifty, I rebased onto master (and fixed the small conflict) and the fix in #481 solved the problem ("Next" button was off the screen, even on my big monitor).  Yay!

Dan points out that this is a good argument for standardizing on running our Selenium tests against Sauce Labs -- monitor size and local setups shouldn't interfere with whether tests pass or fail!

File: psm-app/integration-tests/src/test/java/gov/medicaid/features/enrollment/ui/LicenseInfoPage.java (hunk 3 above)
Comment (jasonaowen) [reply psm-app/integration-tests/src/test/java/gov/medicaid/features/enrollment/ui/LicenseInfoPage.java]: I think this change is in support of the Travis/Saucelabs work, but is a separable change - the phrase "This commit also..." is usually a signal that there are multiple commits that could be split up. I'm fine to disagree and commit on this, though, and I appreciate you updating the commit message.

**Comment by jasonaowen** [comment]
:fireworks: :tada: 

> [The Travis CI build passed](https://travis-ci.org/OpenTechStrategies/psm/builds/294903273)
